### PR TITLE
Add web related natives.

### PR DESCRIPTION
**Pull request by WildCard65**

### CODE REVIEW

**Comment by KyleSanderson** [comment]
I'm a little concerned in regards to your handles, it looks like the pointers are never set from ReadHandle.

In DownloadString you're saving references to the present pawn stack which expires when it leaves scope.

You've left comments in the scripting API questioning what you've done.

Have you tested this?

**Comment by asherkin** [comment]
Nice work!

We're trying to avoid having further blocking APIs though (and are slowly deprecating the ones that exist), it would be especially bad for HTTP given the remote nature. You can look at the Updater extension for an example of using IWebternet from a thread. There also appears to be a lot of debug logging left in.

**Comment by WildCard65** [comment]
Woops must of committed debugging messages when debugging random report
error

**Comment by WildCard65** [comment]
I'll have to thread up this tommorrow.

**Comment by WildCard65** [comment]
Alright, threaded the natives

**Comment by asherkin** [comment]
Have you tested this? It has no synchronization and will almost certainly crash.

**Comment by WildCard65** [comment]
Any suggestions for synchronization?

File: extensions/curl/web_callback.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+DownloadWriteStatus WebCallback::OnDownloadWrite(IWebTransfer *session,
+	void *userdata,
+	void *ptr,
+	size_t size,
+	size_t nmemb)
+{
+	//All but the WebQueueObject call was taken from SM Gamedata Updater MemoryDownloader.cpp
+	char* buffer = nullptr;
+	size_t bufsize = 0, bufpos = 0, total = size * nmemb;
+	if (bufpos + total > bufsize)
```
Comment (KyleSanderson) [inline extensions/curl/web_callback.cpp]: This entire branch will always be hit, and will crash when total == bufsize.

File: extensions/curl/web_thread.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	queue.remove(0);
+}
+
+void WebThread::OnTerminate(IThreadHandle *pHandle, bool cancel)
+{
+	if (!queue.length())
+		return;
+	while (queue.length() > 0)
+	{
+		queue[0].SkipDownload();
```
Comment (KyleSanderson) [inline extensions/curl/web_thread.cpp]: The DBI layer blocks to finish queries, this should do the same.

File: plugins/include/web.inc [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	 * Note: Call this once and only once before DownloadString().
+	 *
+	 * @param form		Handle to web form.
+	 * @return			True on success, false if a form is already set.
+	 * @error			Invalid web session handle.
+	 */
+	public native bool SetPostParam(WebForm form);
+	
+	/*
+	 * Should DownloadString() return false when an http error code is returned?
```
Comment (KyleSanderson) [inline plugins/include/web.inc]: You still have messages questioning the API in the most public facing file possible.

File: plugins/include/web.inc [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	 *
+	 * @param error		True to fail on http error code, false to not.
+	 * @return			True on success, false otherwise.
+	 * @error			Invalid handle.
+	 */
+	public native bool SetErrorOnHttpError(bool error = true);
+	
+	/*
+	 * Should DownloadString() return the http header also?
```
Comment (KyleSanderson) [inline plugins/include/web.inc]: Ditto.

**Comment by KyleSanderson** [comment]
There's too much discussion here to have in a pull request. Please file a bug so we can help you more efficiently with implementing this.

File: extensions/curl/web_callback.cpp (hunk 1 above)
Comment (WildCard65) [inline extensions/curl/web_callback.cpp]: Ok, I'll rework it on Monday.

**Pull request closed without merging**